nullaway/src/main/java/com/uber/nullaway/generics/PreservedAnnotationTreeVisitor.java

```diff
@@ -91,4 +94,99 @@ public Type visitParameterizedType(ParameterizedTypeTree tree, Void p) {
+      return new TypeMetadata(new TypeMetadata.Annotations(attrs));
+    }
+
+    @Override
+    public Type cloneTypeWithMetaData(Type typeToBeCloned, TypeMetadata metaData) {
+      return typeToBeCloned.cloneWithMetadata(metaData);
+    }
+  }
+
+  private static class JDK21TypeMetadataBuilder implements TypeMetadataBuilder {
+
+    private static final MethodHandle typeMetadataHandle = createHandle();
+    private static final MethodHandle cloneAddMetadataHandle =
+        createVirtualMethodHandle(Type.class, TypeMetadata.class, Type.class, "addMetadata");
+    private static final MethodHandle cloneDropMetadataHandle =
+        createVirtualMethodHandle(Type.class, Class.class, Type.class, "dropMetadata");
+
+    private static MethodHandle createHandle() {
+      MethodHandles.Lookup lookup = MethodHandles.lookup();
+      MethodType mt = MethodType.methodType(void.class, com.sun.tools.javac.util.ListBuffer.class);
+      try {
+        return lookup.findConstructor(TypeMetadata.Annotations.class, mt);
+      } catch (NoSuchMethodException e) {
+        throw new RuntimeException(e);
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    private static MethodHandle createVirtualMethodHandle(
+        Class<?> retTypeClass, Class<?> paramTypeClass, Class<?> refClass, String methodName) {
+      MethodHandles.Lookup lookup = MethodHandles.lookup();
+      MethodType mt = MethodType.methodType(retTypeClass, paramTypeClass);
+      try {
+        return lookup.findVirtual(refClass, methodName, mt);
+      } catch (NoSuchMethodException e) {
+        throw new RuntimeException(e);
+      } catch (IllegalAccessException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public TypeMetadata create(com.sun.tools.javac.util.List<Attribute.TypeCompound> attrs) {
+      ListBuffer<Attribute.TypeCompound> b = new ListBuffer<>();
+      b.appendList(attrs);
+      try {
+        return (TypeMetadata) typeMetadataHandle.invoke(b);
+      } catch (Throwable e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    @Override
+    public Type cloneTypeWithMetaData(Type typeToBeCloned, TypeMetadata metaData) {
+      try {
+        // In Jdk21 addMetadata works if there is no metadata associated with the type, so we create
+        // a copy without the existing metadata first and then add it
+        Type clonedTypeWithoutMetadata =
+            (Type) cloneDropMetadataHandle.invoke(typeToBeCloned, metaData.getClass());
+        return (Type) cloneAddMetadataHandle.invoke(clonedTypeWithoutMetadata, metaData);
+      } catch (Throwable e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  private static final TypeMetadataBuilder tmBuilder =
+      getVersion() >= 21
+          ? new JDK21TypeMetadataBuilder()
+          : new JDK17AndEarlierTypeMetadataBuilder();
+
+  private static int getVersion() {
```

Comment:
I think we may want to move this method elsewhere; let me think about a good place for it